In [43]:
import pandas as pd
import numpy as np
from numpy import math

In [44]:
data = pd.read_csv('dataset/raw_dataset.csv')

In [45]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5875 entries, 0 to 5874
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   subject#       5875 non-null   int64  
 1   age            5875 non-null   int64  
 2   sex            5875 non-null   int64  
 3   test_time      5875 non-null   float64
 4   motor_UPDRS    5875 non-null   float64
 5   total_UPDRS    5875 non-null   float64
 6   Jitter(%)      5875 non-null   float64
 7   Jitter(Abs)    5875 non-null   float64
 8   Jitter:RAP     5875 non-null   float64
 9   Jitter:PPQ5    5875 non-null   float64
 10  Jitter:DDP     5875 non-null   float64
 11  Shimmer        5875 non-null   float64
 12  Shimmer(dB)    5875 non-null   float64
 13  Shimmer:APQ3   5875 non-null   float64
 14  Shimmer:APQ5   5875 non-null   float64
 15  Shimmer:APQ11  5875 non-null   float64
 16  Shimmer:DDA    5875 non-null   float64
 17  NHR            5875 non-null   float64
 18  HNR     

This dataset consists of audio medical measurements of 42 people with early-stage Parkinson's disease. These people were hired for a six-month trial of a remote monitoring device to remotely diagnose the disease. Data were automatically recorded in patients' homes.

The columns include an index number of each patient, his age, his gender, the duration from the initial date of admission, two motor indicators UPDRS and total UPDRS, and 16 other medical sound measurements. Each line relates to one of 5,875 different sound measurements from those of the volunteers. Our main goal is to predict the UPDRS (ie 'motor_UPDRS' and 'total_UPDRS') from the 16 sound measurements.

Some more information given to us about the data set are the following:
    

subject - Integer that uniquely identifies each subject

age - Subject age

sex - Subject gender '0' - male, '1' - female

test_time - Time since recruitment into the trial. The integer part is the number of days since recruitment.

motor_UPDRS - Clinician's motor UPDRS score, linearly interpolated

total_UPDRS - Clinician's total UPDRS score, linearly interpolated

Jitter(%),Jitter(Abs),Jitter:RAP,Jitter:PPQ5,Jitter:DDP - Several measures of variation in fundamental frequency

Shimmer,Shimmer(dB),Shimmer:APQ3,Shimmer:APQ5,Shimmer:APQ11,Shimmer:DDA - Several measures of variation in amplitude

NHR,HNR - Two measures of ratio of noise to tonal components in the voice

RPDE - A nonlinear dynamical complexity measure

DFA - Signal fractal scaling exponent

PPE - A nonlinear measure of fundamental frequency variation

In [46]:
data.head()

subject#  age  sex  test_time  motor_UPDRS  total_UPDRS  Jitter(%)  \
0         1   72    0     5.6431       28.199       34.398    0.00662   
1         1   72    0    12.6660       28.447       34.894    0.00300   
2         1   72    0    19.6810       28.695       35.389    0.00481   
3         1   72    0    25.6470       28.905       35.810    0.00528   
4         1   72    0    33.6420       29.187       36.375    0.00335   

   Jitter(Abs)  Jitter:RAP  Jitter:PPQ5  ...  Shimmer(dB)  Shimmer:APQ3  \
0     0.000034     0.00401      0.00317  ...        0.230       0.01438   
1     0.000017     0.00132      0.00150  ...        0.179       0.00994   
2     0.000025     0.00205      0.00208  ...        0.181       0.00734   
3     0.000027     0.00191      0.00264  ...        0.327       0.01106   
4     0.000020     0.00093      0.00130  ...        0.176       0.00679   

   Shimmer:APQ5  Shimmer:APQ11  Shimmer:DDA       NHR     HNR     RPDE  \
0       0.01309        0.01662      0.04314  0.014290  21.640  0.41888   
1       0.01072        0.01689      0.02982  0.011112  27.183  0.43493   
2       0.00844        0.01458      0.02202  0.020220  23.047  0.46222   
3       0.01265        0.01963      0.03317  0.027837  24.445  0.48730   
4       0.00929        0.01819      0.02036  0.011625  26.126  0.47188   

       DFA      PPE  
0  0.54842  0.16006  
1  0.56477  0.10810  
2  0.54405  0.21014  
3  0.57794  0.33277  
4  0.56122  0.19361  

[5 rows x 22 columns]

In [47]:
data.drop(['subject#'],axis=1,inplace=True)

We see that in our data set there is a feature called subject #, which basically shows us which patient each one is. Its layout, however, does not make sense (it is unordered), so it can then confuse the training of our classifiers. So we need to find a way to handle it so that it gives us information that makes sense.

We can either convert this feature with one-hot encoding or discard it completely. The first method will add too many features, so we will try the second.





Before doing anything, we must make sure that there are no missing values ​​in various features. This is an important step, because if there are any we have to decide how to handle them. We can say we delete them either if we follow other techniques such as those of Expectation Maximization (EM) or pseudo-EM.

In our case fortunately there are no missing values ​​at all, so we will not worry about this problem.


In [48]:
data.isnull().sum()

age              0
sex              0
test_time        0
motor_UPDRS      0
total_UPDRS      0
Jitter(%)        0
Jitter(Abs)      0
Jitter:RAP       0
Jitter:PPQ5      0
Jitter:DDP       0
Shimmer          0
Shimmer(dB)      0
Shimmer:APQ3     0
Shimmer:APQ5     0
Shimmer:APQ11    0
Shimmer:DDA      0
NHR              0
HNR              0
RPDE             0
DFA              0
PPE              0
dtype: int64

According to our work we have to divide dataset into 4 categories and work upon 4 individually


In [49]:
motor_16_data = data.copy()
total_16_data = data.copy()
motor_18_data = data.copy()
total_18_data = data.copy()

We have 4 datasets from raw datasets:
    1 -- All features to predict motorUPDRS
    2 -- Drop features (Age And Sex ) to predict motorUPDRS
    3 -- All features to predict totalUPDRS
    4 -- Drop features (Age And Sex ) to predict totalUPDRS

In [50]:
def drop_column(x):
    if(x==0):
        dataset = motor_16_data
        motor_16_data.drop(['age','sex'],axis=1,inplace=True)
    elif(x==1):
        dataset = total_16_data
        total_16_data.drop(['age','sex'],axis=1,inplace=True)

In [51]:
for i in range(2):
    drop_column(i)

In [52]:
motor_16_data.to_csv("motor_16_data.csv",index=False)
total_16_data.to_csv("total_16_data.csv",index=False)
motor_18_data.to_csv("motor_18_data.csv",index=False)
total_18_data.to_csv("total_18_data.csv",index=False)

In [53]:
m16 = pd.read_csv('motor_16_data.csv')
t16 = pd.read_csv('total_16_data.csv')
m18 = pd.read_csv('motor_18_data.csv')
t18 = pd.read_csv('total_18_data.csv')

Now we Analyse all the datsets and use the Algorithm according to that.

In [54]:
m16.describe()

test_time  motor_UPDRS  total_UPDRS    Jitter(%)  Jitter(Abs)  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000   
mean     92.863722    21.296229    29.018942     0.006154     0.000044   
std      53.445602     8.129282    10.700283     0.005624     0.000036   
min      -4.262500     5.037700     7.000000     0.000830     0.000002   
25%      46.847500    15.000000    21.371000     0.003580     0.000022   
50%      91.523000    20.871000    27.576000     0.004900     0.000035   
75%     138.445000    27.596500    36.399000     0.006800     0.000053   
max     215.490000    39.511000    54.992000     0.099990     0.000446   

        Jitter:RAP  Jitter:PPQ5   Jitter:DDP      Shimmer  Shimmer(dB)  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000   
mean      0.002987     0.003277     0.008962     0.034035     0.310960   
std       0.003124     0.003732     0.009371     0.025835     0.230254   
min       0.000330     0.000430     0.000980     0.003060     0.026000   
25%       0.001580     0.001820     0.004730     0.019120     0.175000   
50%       0.002250     0.002490     0.006750     0.027510     0.253000   
75%       0.003290     0.003460     0.009870     0.039750     0.365000   
max       0.057540     0.069560     0.172630     0.268630     2.107000   

       Shimmer:APQ3  Shimmer:APQ5  Shimmer:APQ11  Shimmer:DDA          NHR  \
count   5875.000000   5875.000000    5875.000000  5875.000000  5875.000000   
mean       0.017156      0.020144       0.027481     0.051467     0.032120   
std        0.013237      0.016664       0.019986     0.039711     0.059692   
min        0.001610      0.001940       0.002490     0.004840     0.000286   
25%        0.009280      0.010790       0.015665     0.027830     0.010955   
50%        0.013700      0.015940       0.022710     0.041110     0.018448   
75%        0.020575      0.023755       0.032715     0.061735     0.031463   
max        0.162670      0.167020       0.275460     0.488020     0.748260   

               HNR         RPDE          DFA          PPE  
count  5875.000000  5875.000000  5875.000000  5875.000000  
mean     21.679495     0.541473     0.653240     0.219589  
std       4.291096     0.100986     0.070902     0.091498  
min       1.659000     0.151020     0.514040     0.021983  
25%      19.406000     0.469785     0.596180     0.156340  
50%      21.920000     0.542250     0.643600     0.205500  
75%      24.444000     0.614045     0.711335     0.264490  
max      37.875000     0.966080     0.865600     0.731730

In [55]:
m18.describe()

age          sex    test_time  motor_UPDRS  total_UPDRS  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000   
mean     64.804936     0.317787    92.863722    21.296229    29.018942   
std       8.821524     0.465656    53.445602     8.129282    10.700283   
min      36.000000     0.000000    -4.262500     5.037700     7.000000   
25%      58.000000     0.000000    46.847500    15.000000    21.371000   
50%      65.000000     0.000000    91.523000    20.871000    27.576000   
75%      72.000000     1.000000   138.445000    27.596500    36.399000   
max      85.000000     1.000000   215.490000    39.511000    54.992000   

         Jitter(%)  Jitter(Abs)   Jitter:RAP  Jitter:PPQ5   Jitter:DDP  ...  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000  ...   
mean      0.006154     0.000044     0.002987     0.003277     0.008962  ...   
std       0.005624     0.000036     0.003124     0.003732     0.009371  ...   
min       0.000830     0.000002     0.000330     0.000430     0.000980  ...   
25%       0.003580     0.000022     0.001580     0.001820     0.004730  ...   
50%       0.004900     0.000035     0.002250     0.002490     0.006750  ...   
75%       0.006800     0.000053     0.003290     0.003460     0.009870  ...   
max       0.099990     0.000446     0.057540     0.069560     0.172630  ...   

       Shimmer(dB)  Shimmer:APQ3  Shimmer:APQ5  Shimmer:APQ11  Shimmer:DDA  \
count  5875.000000   5875.000000   5875.000000    5875.000000  5875.000000   
mean      0.310960      0.017156      0.020144       0.027481     0.051467   
std       0.230254      0.013237      0.016664       0.019986     0.039711   
min       0.026000      0.001610      0.001940       0.002490     0.004840   
25%       0.175000      0.009280      0.010790       0.015665     0.027830   
50%       0.253000      0.013700      0.015940       0.022710     0.041110   
75%       0.365000      0.020575      0.023755       0.032715     0.061735   
max       2.107000      0.162670      0.167020       0.275460     0.488020   

               NHR          HNR         RPDE          DFA          PPE  
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000  
mean      0.032120    21.679495     0.541473     0.653240     0.219589  
std       0.059692     4.291096     0.100986     0.070902     0.091498  
min       0.000286     1.659000     0.151020     0.514040     0.021983  
25%       0.010955    19.406000     0.469785     0.596180     0.156340  
50%       0.018448    21.920000     0.542250     0.643600     0.205500  
75%       0.031463    24.444000     0.614045     0.711335     0.264490  
max       0.748260    37.875000     0.966080     0.865600     0.731730  

[8 rows x 21 columns]

In [56]:
t16.describe()

test_time  motor_UPDRS  total_UPDRS    Jitter(%)  Jitter(Abs)  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000   
mean     92.863722    21.296229    29.018942     0.006154     0.000044   
std      53.445602     8.129282    10.700283     0.005624     0.000036   
min      -4.262500     5.037700     7.000000     0.000830     0.000002   
25%      46.847500    15.000000    21.371000     0.003580     0.000022   
50%      91.523000    20.871000    27.576000     0.004900     0.000035   
75%     138.445000    27.596500    36.399000     0.006800     0.000053   
max     215.490000    39.511000    54.992000     0.099990     0.000446   

        Jitter:RAP  Jitter:PPQ5   Jitter:DDP      Shimmer  Shimmer(dB)  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000   
mean      0.002987     0.003277     0.008962     0.034035     0.310960   
std       0.003124     0.003732     0.009371     0.025835     0.230254   
min       0.000330     0.000430     0.000980     0.003060     0.026000   
25%       0.001580     0.001820     0.004730     0.019120     0.175000   
50%       0.002250     0.002490     0.006750     0.027510     0.253000   
75%       0.003290     0.003460     0.009870     0.039750     0.365000   
max       0.057540     0.069560     0.172630     0.268630     2.107000   

       Shimmer:APQ3  Shimmer:APQ5  Shimmer:APQ11  Shimmer:DDA          NHR  \
count   5875.000000   5875.000000    5875.000000  5875.000000  5875.000000   
mean       0.017156      0.020144       0.027481     0.051467     0.032120   
std        0.013237      0.016664       0.019986     0.039711     0.059692   
min        0.001610      0.001940       0.002490     0.004840     0.000286   
25%        0.009280      0.010790       0.015665     0.027830     0.010955   
50%        0.013700      0.015940       0.022710     0.041110     0.018448   
75%        0.020575      0.023755       0.032715     0.061735     0.031463   
max        0.162670      0.167020       0.275460     0.488020     0.748260   

               HNR         RPDE          DFA          PPE  
count  5875.000000  5875.000000  5875.000000  5875.000000  
mean     21.679495     0.541473     0.653240     0.219589  
std       4.291096     0.100986     0.070902     0.091498  
min       1.659000     0.151020     0.514040     0.021983  
25%      19.406000     0.469785     0.596180     0.156340  
50%      21.920000     0.542250     0.643600     0.205500  
75%      24.444000     0.614045     0.711335     0.264490  
max      37.875000     0.966080     0.865600     0.731730

In [57]:
t18.describe()

age          sex    test_time  motor_UPDRS  total_UPDRS  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000   
mean     64.804936     0.317787    92.863722    21.296229    29.018942   
std       8.821524     0.465656    53.445602     8.129282    10.700283   
min      36.000000     0.000000    -4.262500     5.037700     7.000000   
25%      58.000000     0.000000    46.847500    15.000000    21.371000   
50%      65.000000     0.000000    91.523000    20.871000    27.576000   
75%      72.000000     1.000000   138.445000    27.596500    36.399000   
max      85.000000     1.000000   215.490000    39.511000    54.992000   

         Jitter(%)  Jitter(Abs)   Jitter:RAP  Jitter:PPQ5   Jitter:DDP  ...  \
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000  ...   
mean      0.006154     0.000044     0.002987     0.003277     0.008962  ...   
std       0.005624     0.000036     0.003124     0.003732     0.009371  ...   
min       0.000830     0.000002     0.000330     0.000430     0.000980  ...   
25%       0.003580     0.000022     0.001580     0.001820     0.004730  ...   
50%       0.004900     0.000035     0.002250     0.002490     0.006750  ...   
75%       0.006800     0.000053     0.003290     0.003460     0.009870  ...   
max       0.099990     0.000446     0.057540     0.069560     0.172630  ...   

       Shimmer(dB)  Shimmer:APQ3  Shimmer:APQ5  Shimmer:APQ11  Shimmer:DDA  \
count  5875.000000   5875.000000   5875.000000    5875.000000  5875.000000   
mean      0.310960      0.017156      0.020144       0.027481     0.051467   
std       0.230254      0.013237      0.016664       0.019986     0.039711   
min       0.026000      0.001610      0.001940       0.002490     0.004840   
25%       0.175000      0.009280      0.010790       0.015665     0.027830   
50%       0.253000      0.013700      0.015940       0.022710     0.041110   
75%       0.365000      0.020575      0.023755       0.032715     0.061735   
max       2.107000      0.162670      0.167020       0.275460     0.488020   

               NHR          HNR         RPDE          DFA          PPE  
count  5875.000000  5875.000000  5875.000000  5875.000000  5875.000000  
mean      0.032120    21.679495     0.541473     0.653240     0.219589  
std       0.059692     4.291096     0.100986     0.070902     0.091498  
min       0.000286     1.659000     0.151020     0.514040     0.021983  
25%       0.010955    19.406000     0.469785     0.596180     0.156340  
50%       0.018448    21.920000     0.542250     0.643600     0.205500  
75%       0.031463    24.444000     0.614045     0.711335     0.264490  
max       0.748260    37.875000     0.966080     0.865600     0.731730  

[8 rows x 21 columns]